src/normalize/kernel.rs

```diff
@@ -0,0 +1,248 @@
+            // It's unclear whether we should error out here or map anything
+            // "unknown" to `Unknown`.
+            x => Err(Error::with_invalid_data(format!(
+                "{PROC_RANDOMIZE_VA_SPACE} node value {x} is not understood"
+            ))),
+        }
+    }
+}
+
+
+/// # Notes
+/// Right now this function imposes an arbitrary limit on the maximum
+/// node value content size.
+fn read_proc_node_value<T>(path: &Path) -> Result<Option<T>>
+where
+    T: FromStr,
+    T::Err: StdError + Send + Sync + 'static,
+{
+    let result = File::open(path);
+    let mut file = match result {
+        Ok(file) => file,
+        Err(err) if err.kind() == io::ErrorKind::NotFound => return Ok(None),
+        Err(err) => return Err(err.into()),
+    };
+
+    // We don't want to blindly use Read::read_to_end or something like
+    // that if we can avoid it.
+    let mut buffer = [0; u8::MAX as usize];
+    let count = file.read(&mut buffer)?;
+    if count >= size_of_val(&buffer) {
+        return Err(Error::with_invalid_data(format!(
+            "file content is larger than {} bytes",
+            size_of_val(&buffer)
+        )))
+    }
+
+    let s = str::from_utf8(&buffer[0..count]).map_err(Error::with_invalid_data)?;
+    let value = T::from_str(s).map_err(Error::with_invalid_data)?;
+    Ok(Some(value))
+}
+
+
+/// Try to determine the KASLR state of the system.
+fn determine_kaslr_state() -> Result<KaslrState> {
+    // https://www.kernel.org/doc/html/latest/admin-guide/sysctl/kernel.html#randomize-va-space
+    let kaslr = read_proc_node_value::<KaslrState>(Path::new(PROC_RANDOMIZE_VA_SPACE))
+        .with_context(|| {
+            format!(
+                "failed to determine KASLR state from {}",
+                PROC_RANDOMIZE_VA_SPACE
+            )
+        })?
+        .unwrap_or(KaslrState::Unknown);
+    Ok(kaslr)
+}
+
+/// "Parse" the VMCOREINFO descriptor.
+///
+/// This underspecified blob roughly has the following format:
+/// ```
+/// OSRELEASE=6.2.15-100.fc36.x86_64
+/// BUILD-ID=d3d01c80278f8927486b7f01d0ab6be77784dceb
+/// PAGESIZE=4096
+/// SYMBOL(init_uts_ns)=ffffffffb72b8160
+/// OFFSET(uts_namespace.name)=0
+/// [...]
+/// ```
+fn parse_vmcoreinfo_desc(desc: &[u8]) -> impl Iterator<Item = (&[u8], &[u8])> {
+    desc.split(|&b| b == b'\n')
+        .filter_map(|line| split_bytes(line, |b| b == b'='))
+}
+
+/// Find and read the `KERNELOFFSET` note in a "kcore" file represented by
+/// `parser` (i.e., already opened as an ELF).
+fn find_kaslr_offset(parser: &ElfParser<File>) -> Result<Option<u64>> {
+    let phdrs = parser.program_headers()?;
+    for phdr in phdrs.iter(0) {
+        if phdr.type_() != elf::types::PT_NOTE {
+            continue
+        }
+
+        let file = parser.backend();
+        let mut offset = phdr.offset();
+
+        // Iterate through all available notes. See `elf(5)` for
+        // details.
+        while offset + (size_of::<ElfN_Nhdr>() as u64) <= phdr.file_size() {
```

Comment:
nit: blazesym parses Elf notes at least in two places now, right? Why not add a proper iterator support and clean up this code (and the one that does build ID, right?)

Reply:
No big deal and can be done as a follow up. But I do find all those offset checks, file size, etc quite mundane and error prone, so I still feel like iterator would be an improvement. It can return Nhdr + name/descr slices, which would save a bunch of error check and otherwise distracting plumbing. But up to you.